scripts/zephyr_module.py

```diff
@@ -0,0 +1,145 @@
+#!/usr/bin/env python3
+#
+# Copyright (c) 2019, Nordic Semiconductor ASA
+#
+# SPDX-License-Identifier: Apache-2.0
+
+'''Tool for parsing a list of projects to determine if they are Zephyr
+projects. If no projects are given then the output from `west list` will be
+used as project list.
+
+Include file is generated for Kconfig using --kconfig-out.
+A <name>:<path> text file is generated for use with CMake using --cmake-out.
+'''
+
+import argparse
+import os
+import sys
+import yaml
+import pykwalify.core
+import subprocess
+import re
+
+
+METADATA_SCHEMA = '''\
+## A pykwalify schema for basic validation of the structure of a
+## metadata YAML file.
+##
+# The zephyr/module.yml file is a simple list of key value pairs to be used by
+# the build system.
+type: map
+mapping:
+  build:
+    required: true
+    type: map
+    mapping:
+      cmake:
+        required: false
+        type: str
+      kconfig:
+        required: false
+        type: str
+'''
+
+AUTO_GENERATED_HEADER = '# NOTE: THIS FILE IS AUTOGENERATED - DO NOT EDIT\n\n'
+
+
+def process_module(module, section_name, cmake_out=None, kconfig_out=None):
+    cmake_path = 'zephyr'
+    kconfig_file = 'zephyr/Kconfig'
+
+    module_yml = os.path.join(module, 'zephyr/module.yml')
+    if os.path.isfile(module_yml):
+        with open(module_yml, 'r') as f:
+            meta = yaml.safe_load(f.read())
+
+        schema = yaml.safe_load(METADATA_SCHEMA)
+        pykwalify.core.Core(source_data=meta, schema_data=schema).validate()
+
+        section = meta.get(section_name, dict())
+        if section is not None:
+            cmake_path = section.get('cmake', cmake_path)
+            kconfig_file = section.get('kconfig', kconfig_file)
+
+    cmake_path_abs = os.path.abspath(os.path.join(module, cmake_path))
+    cmake_file_abs = os.path.join(cmake_path_abs, 'CMakeLists.txt')
+    if os.path.isfile(cmake_file_abs) and cmake_out is not None:
+        val_str = '{}:{}\n'.format(os.path.basename(module), cmake_path_abs)
+
+        with open(cmake_out, 'a') as f:
+            f.write(val_str)
+
+    kconfig_file_abs = os.path.abspath(os.path.join(module, kconfig_file))
+    if os.path.isfile(kconfig_file_abs) and kconfig_out is not None:
+        val_str = 'osource "{}"\n\n'.format(kconfig_file_abs)
+
+        with open(kconfig_out, 'a') as f:
+            f.write(val_str)
+
+
+def generate_file_header(includefile):
+    path = os.path.dirname(includefile)
+    if path != '' and not os.path.exists(path):
+        os.makedirs(path)
```

Comment:
This can be dropped, the user should be able to make the directory of the path he inputs.

Reply:
Well, the `user` in this case is CMake, and at this point in time, then the folder `${PROJECT_BINARY_DIR}` doesn't exist.
The reason this worked in CMake was because `file(write <path>/filename)` will automatically create `<path>`, if it doesn't exists.

`${PROJECT_BINARY_DIR}` is used because that variable is already exported to Kconfig

Reply:
CMake should create the directory then.

Reply:
Folder creation moved to CMake.